graph_djikstra_heap.c

```diff
@@ -0,0 +1,240 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#define HEAP_MAX_SIZE 100
+#define HEAP_INIT_SIZE 10
+#define P_QUEUE_ELEMENT_SIZE 2
+
+/*
+pq based on min binary heap, e.g. 10,20,50,30,40,60,70
+heap keeps array of arrays where each element is int* array[2] = [priority, key]
+*/
+
+typedef struct {
+    int **p_queue;
+    int size;
+} Heap;
+
+Heap* heap_create();
+void heap_free(Heap *);
+
+void heap_push(Heap *, int *);
+int* heap_peek(Heap *);
+int* heap_pop(Heap *);
+void heap_print(Heap *);
+
+static void heap_realloc(Heap *);
+static void heap_heapify(Heap *);
+
+void heap_print(Heap* self) {
+    for(int i = 0; i < self->size; i++) {
+        printf("%d ", self->p_queue[i][0]);
+        printf("%d\n", self->p_queue[i][1]);
+    }
+}
+
+int* heap_peek(Heap *self) {
+    return self->p_queue[0];
+}
+
+int* heap_pop(Heap *self) {
+    if (self->size <=0) {
+        perror("Heap is empty");
+        abort();
+    }
+
+    int* popped_arr = self->p_queue[0];
+    self->size--;
+
+    if (self->size <= 0) {
+        return popped_arr;
+    }
+
+    self->p_queue[0] = self->p_queue[self->size];
```

Comment:
self->size - 1?

Reply:
Decreased on line 47
You are right as it may be not so clear

Reply:
its fine mr. Sir. Next time i have to be more careful